python/pyarrow/public-api.pxi

```diff
@@ -92,6 +92,31 @@ cdef public api object pyarrow_wrap_data_type(
     return out
 
 
+cdef object pyarrow_wrap_metadata(
+        const shared_ptr[const CKeyValueMetadata]& meta):
+    cdef const CKeyValueMetadata* cmeta = meta.get()
+
+    if cmeta == nullptr:
+        return None
+
+    result = OrderedDict()
+    for i in range(cmeta.size()):
+        result[cmeta.key(i)] = cmeta.value(i)
+
+    return result
+
+
+cdef shared_ptr[CKeyValueMetadata] pyarrow_unwrap_metadata(object meta):
+    cdef vector[c_string] keys, values
+
+    if isinstance(meta, dict):
+        keys = map(tobytes, meta.keys())
+        values = map(tobytes, meta.values())
+        return make_shared[CKeyValueMetadata](keys, values)
+
+    return shared_ptr[CKeyValueMetadata]()
```

Comment:
I know, but the same is true for all of the `pyarrow_unwrap_*` functions, so this is really just because of consistency. We should handle invalid arguments better in all unwrap functions.